awsx/ec2/subnetDistributorNew.ts

```diff
@@ -0,0 +1,296 @@
+ * @param {number} n input number
+ * @returns {number} next power of 2 to n (>= n)
+ */
+function nextPow2(n: number): number {
+  if (n === 0) {
+    return 1;
+  }
+
+  n--;
+  n |= n >> 1;
+  n |= n >> 2;
+  n |= n >> 4;
+  n |= n >> 8;
+  n |= n >> 16;
+
+  return n + 1;
+}
+
+export type ExplicitSubnetSpecInputs = SubnetSpecInputs &
+  Required<Pick<SubnetSpecInputs, "cidrBlocks">>;
+
+export function getSubnetSpecsExplicit(
+  vpcName: string,
+  azNames: string[],
+  subnetInputs: ExplicitSubnetSpecInputs[],
+): SubnetSpec[] {
+  const subnets: SubnetSpec[] = [];
+  for (let azIndex = 0; azIndex < azNames.length; azIndex++) {
+    const azName = azNames[azIndex];
+    const azNum = azIndex + 1;
+    for (const subnetSpec of subnetInputs) {
+      const subnetCidr = subnetSpec.cidrBlocks[azIndex];
+      subnets.push({
+        cidrBlock: subnetCidr,
+        type: subnetSpec.type,
+        azName,
+        subnetName: subnetName(vpcName, subnetSpec, azNum),
+      });
+    }
+  }
+
+  return subnets;
+}
+
+function subnetName(vpcName: string, subnet: SubnetSpecInputs, azNum: number): string {
+  const specName = subnet.name ?? subnet.type.toLowerCase();
+  return `${vpcName}-${specName}-${azNum}`;
+}
+
+/* Ensure all inputs are consistent and fill in missing values with defaults
+ * Ensure any specified, netmask, size or blocks are in agreement.
+ */
+export function validateAndNormalizeSubnetInputs(
+  subnetArgs: SubnetSpecInputs[] | undefined,
+  availabilityZoneCount: number,
+):
+  | { normalizedSpecs: SubnetSpecInputs[]; isExplicitLayout: false }
+  | { normalizedSpecs: ExplicitSubnetSpecInputs[]; isExplicitLayout: true }
+  | undefined {
+  if (subnetArgs === undefined) {
+    return undefined;
+  }
+
+  const issues: string[] = [];
+
+  // All sizes must be valid.
+  const invalidSizes = subnetArgs.filter(
+    (spec) => spec.size !== undefined && !validSubnetSizes.includes(spec.size),
+  );
+  if (invalidSizes.length > 0) {
+    issues.push(
+      `The following subnet sizes are invalid: ${invalidSizes
+        .map((spec) => spec.size)
+        .join(", ")}. Valid sizes are: ${validSubnetSizes.join(", ")}.`,
+    );
+  }
+
+  const hasExplicitLayouts = subnetArgs.some((subnet) => subnet.cidrBlocks !== undefined);
+  if (hasExplicitLayouts) {
+    const explicitSpecs: ExplicitSubnetSpecInputs[] = [];
+
+    for (let specIndex = 0; specIndex < subnetArgs.length; specIndex++) {
+      const spec = subnetArgs[specIndex];
+      const cidrBlocks = spec.cidrBlocks;
+
+      // If any subnet spec has explicit cidrBlocks, all subnets must have explicit cidrBlocks.
+      if (cidrBlocks === undefined) {
+        issues.push(
+          "If any subnet spec has explicit cidrBlocks, all subnets must have explicit cidrBlocks.",
+        );
+        continue;
+      }
+
+      // Number of cidrBlocks must match the number of availability zones.
+      if (cidrBlocks !== undefined && cidrBlocks.length !== availabilityZoneCount) {
```

Comment:
Aren't you already checking if cidrBlocks is undefined above?

Reply:
Ah yes! I missed that bit in the refactor 🙇